### PR TITLE
[Feature] Use community relation of search requests

### LINKED ISSUE #10366: ✨ Update Request pages to include Community scope
## ✨ Feature

#10331 ensured every search request is associated with a community, but didn't do anything with that data. Now, we want to
1. explicitly scope the search results by community
2. label the community on the Request page, on the admin side
3. ensure new Community roles can see requests to their community

## 🕵️ Details

1. Community field in ApplicantFilter scopes search results to candidates in pools in that community
2. Somewhere on the Single Request page, the associated Community is listed
3. Community becomes an input field in Candidate Table filter modal
4. Community column added to Request table, but hidden by default
5. Requests index and Single Request page become available to Community Admin and Community Recruiter roles. Link to Requests page is available on dashboard and in nav menu.
6.  authorizedToView scope on requests query should ensure the new roles only see requests in their communities. Policies to view requests, to search candidates, and to update requests (notes, status) are updated for these roles, if necessary. Legacy roles continue to work.

## 🎨 Design File

[Figma file](https://www.figma.com/design/em5ArPr044YaIe1Vb1zCZz/Record-of-Decision-(Communities)?node-id=9374-59294&t=2MFpsy3hqotDbLKI-1) by @Jerryescandon 

## 📸 Screenshot

Jerbo: "Given the amount of fields, we might need to rework this whole card into 3 columns instead"

image

## 🙋‍♀️ Proposed Implementation

(optional) Do you have a proposed implementation?

## 🌎 Localization

(optional) Provide any new copy along with translations available.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] Community field in ApplicantFilter scopes search results to candidates in pools in that community
- [ ] Somewhere on the Single Request page, the associated Community is listed
- [ ] Community becomes an input field in PoolCandidates Table filter modal
- [ ] Community column added to Request table, but hidden by default
- [ ] Requests index and Single Request page become available to Community Admin and Community Recruiter roles. 
- [ ] authorizedToView scope on requests query should ensure the new roles only see requests in their communities. Policies to view requests, to search candidates, and to update requests (notes, status) are updated for these roles, if necessary. Legacy roles continue to work.
- [ ] phpunit tests that the community scope works for candidate search, and that the new roles can only see the correct requests and candidates
- [ ] playwright tests for new roles able to interact with these pages

## 🛑 Blockers

Issues which must be completed before this one.

```[tasklist]
### Blocked By
- [ ] #10302 
- [ ] #10331 
- [ ] https://github.com/GCTC-NTGC/gc-digital-talent/issues/10364
```

### PR DESCRIPTION
**Pull request by vd1992**
🤖 Resolves #10366

## 👋 Introduction

Does quite a few things! Might be best to just look over the issue first. 

## 🕵️ Details

Recruiter is recruiter@test.com
Admin is community@test.com

## 🧪 Testing

1. Login as a new community role, so admin or recruiter
2. Observe requests table
3. View a request and observe community, test mutations
4. Can see candidates in the table
5. Login as super admin
6. Filter candidates table by community

## 📸 Screenshot

image

image

image

image

image

image

## 🚚 Deployment

Run the role permission seeder

### CODE REVIEW

File: Makefile
Comment (vd1992) [inline Makefile:1]: Fine tuned once I understood how the command worked

`make test` for all
or
`make test CMD="--filter abc"` to filter conveniently

File: api/database/seeders/SearchRequestRandomSeeder.php
Comment (vd1992) [inline api/database/seeders/SearchRequestRandomSeeder.php:1]: I think this is now the intent and not the prior state of 50 requests attached to a single applicant filter

File: apps/playwright/tests/admin/auth.spec.ts
Comment (vd1992) [inline apps/playwright/tests/admin/auth.spec.ts:1]: Definitely not the end state, but matching up the roles to everything they should be able to see will come later

File: api/app/Models/Pool.php
Comment (vd1992) [inline api/app/Models/Pool.php:1]: Needing to add to `$selectable` tripped me up for a couple of hours 😞

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Looking good!

File: api/database/seeders/SearchRequestRandomSeeder.php
Comment (vd1992) [reply api/database/seeders/SearchRequestRandomSeeder.php:1]: Yeah sure, completed over the course of
https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/ee16dc637466cb5e2cdad8e2afb4bd4c0e609bdb
https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/8f895ac7eb5d59c99838ac89fb0d54d93fea8f42

File: api/app/Models/PoolCandidate.php [hunk 1]
```diff
@@ -295,6 +295,22 @@ public static function scopeQualifiedStreams(Builder $query, ?array $streams): B
...
+     */
+    public static function scopeCandidatesInCommunity(Builder $query, ?string $communityId): Builder
+    {
+        if (empty($communityId)) {
+            return $query;
+        }
+
+        $query->whereHas('pool.community', function ($query) use ($communityId) {
+            $query->where('id', $communityId);
+        });
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/PoolCandidate.php]: Yep, makes sense to me, done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/8cbef0901856e209306779ea5cec37143dab2a8f

File: api/app/Models/Pool.php [hunk 2]
```diff
@@ -127,6 +127,7 @@ class Pool extends Model
         'closing_reason',
         'process_number',
         'department_id',
+        'community_id',
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/Pool.php:131]: https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/e9a3b2063a598b835352e3ccfbf44364efb2fe54

File: api/app/Models/PoolCandidateSearchRequest.php [hunk 3]
```diff
@@ -242,6 +244,43 @@ public static function scopeGeneralSearch(Builder $query, ?string $search): Buil
...
 
+    /**
+     * Scope the query to SearchRequests the current user can view
+     */
+    public function scopeAuthorizedToView(Builder $query)
+    {
+        /** @var \App\Models\User */
+        $user = Auth::user();
+
+        if (! $user) {
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/PoolCandidateSearchRequest.php]: Yeah sure, done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/7e3ee9174d44d125806e1b02a7b554adfdca7fd6

File: api/database/factories/CommunityFactory.php [hunk 4]
```diff
@@ -31,6 +31,15 @@ public function definition()
         ];
     }
 
+    public function configure()
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/database/factories/CommunityFactory.php]: Oh okay, I thought we wanted them created by default and it was missing
Reversed in https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/8ce51cbbe75e3c69ceade956f12d573311b52200

File: api/database/factories/PoolFactory.php [hunk 5]
```diff
@@ -91,6 +91,15 @@ public function definition()
         ];
     }
 
+    public function configure()
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/database/factories/PoolFactory.php]: https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/8ce51cbbe75e3c69ceade956f12d573311b52200

File: apps/playwright/fixtures/CommunityRecruiterPage.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    await this.graphqlRequest(Test_UpdateUserRolesMutationDocument, {
+      updateUserRolesInput: {
+        userId,
+        roleAssignmentsInput: {
+          attach: roleInputArray,
+        },
+      },
+    });
+  }
```
[the remark this answers is not in the document]
Comment (vd1992) [reply apps/playwright/fixtures/CommunityRecruiterPage.ts:42]: 🤷‍♂️ 
I am just trying to keep things consistent with what already exists, so I dunno I just copied what was already there

[the remark this answers is not in the document]
Comment (tristan-orourke) [reply apps/playwright/fixtures/CommunityRecruiterPage.ts:42]: Yeah, this is duplicated now across  three different classes. Let's move `createUser` and `addRolesToUser` into the AppPage class.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Thanks, just two small changes left I'd like to see.

File: apps/playwright/fixtures/CommunityRecruiterPage.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vd1992) [reply apps/playwright/fixtures/CommunityRecruiterPage.ts]: Sure
https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/a3c5ce8ab8f37b98b7add4a2037a48ce30f3eddc
https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/17dadd9c0ca1f531fd5475553f26326a7cd3771e
Turns out I didn't need the change in the second commit

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Just a couple comments about changes to tables:
- If you show the Community column in the Talent Requests table, it appears to be sortable, but clicking it doesn't change any ordering
- If you go to a View Talent Request page and open the filter dialogs, Community is always empty. It should be initialized with the community from the request filter.

**Comment by vd1992** [comment]
Ah, thanks 

1. Yeah, I overlooked that. Removed the sortability so it is like the rest of the columns with chained relations https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/3c2c90f49cb5cacb0938a3e8f7db1007b69e9ecf
2. Oh, whoops. A little tricky to get to the bottom of, but think I got it in https://github.com/GCTC-NTGC/gc-digital-talent/pull/11132/commits/27791bcf7c46f36378e0e6c44f9bcf7e153a068a

**Review by tristan-orourke — APPROVED** [review]
All good!

**Pull request merged**